polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -128,48 +137,50 @@ class PolyKernel private[kernel] (
-                        statusUpdates.publish1(UpdatedTasks(taskInfo.copy(detail = newDetail, progress = (progress * 255).toByte) :: Nil)).unsafeRunSync()
-                    }
-
-                    symbolTable.drain() *> interp.runCode(
-                      id,
-                      findAvailableSymbols(prevCellIds, interp),
-                      prevCellIds,
-                      cell.content.toString
-                    ).map {
-                      results => predefResults merge results.evalMap {
-                        case v: ResultValue =>
-                          symbolTable.publishAll(symbolTable.RuntimeValue.fromResultValue(v, interp).toList).as(v)
-                        case result =>
-                          IO.pure(result)
-                      }
-                    }.handleErrorWith {
-                      case errs@CompileErrors(_) => IO.pure(Stream.emit(errs))
-                      case err@RuntimeError(_) => IO.pure(Stream.emit(err))
-                      case err => IO.pure(Stream.emit(RuntimeError(err)))
-                    }.guarantee(oq.enqueue1(None)) <* symbolTable.drain()
+                // TODO: should this be something the interpreter has to do? Without this we wouldn't even need to allocate a queue here
+                polynote.runtime.Runtime.setDisplayer((mime, content) => oqSome.enqueue1(Output(mime, content)).unsafeRunSync())
+                polynote.runtime.Runtime.setProgressSetter {
+                  (progress, detail) =>
+                    val newDetail = Option(detail).getOrElse(taskInfo.detail)
+                    statusUpdates.publish1(UpdatedTasks(taskInfo.copy(detail = newDetail, progress = (progress * 255).toByte) :: Nil)).unsafeRunSync()
                 }
-            }.handleErrorWith(err => IO.pure(Stream.emit(RuntimeError(err))))
-          }
-        Stream.emit(ClearResults()) ++ Stream(cellResults.flatten, oq.dequeue.unNoneTerminate).parJoinUnbounded
+
+                symbolTable.drain() *> interp.runCode(
+                  id,
+                  findAvailableSymbols(prevCellIds, interp),
+                  prevCellIds,
+                  cell.content.toString
+                ).map {
+                  results => results.evalMap {
+                    case v: ResultValue =>
+                      symbolTable.publishAll(symbolTable.RuntimeValue.fromResultValue(v, interp).toList).as(v)
+                    case result =>
+                      IO.pure(result)
+                  }
+                }.handleErrorWith {
+                  case errs@CompileErrors(_) => IO.pure(Stream.emit(errs))
+                  case err@RuntimeError(_) => IO.pure(Stream.emit(err))
+                  case err => IO.pure(Stream.emit(RuntimeError(err)))
+                }
+            }.map(_.getOrElse(Stream.empty)).handleErrorWith(err => IO.pure(Stream.emit(RuntimeError(err)))).map {
+              _ ++ Stream.eval(oq.enqueue1(None)).drain ++ Stream.eval(symbolTable.drain()).drain
```

Comment:
Can you explain a bit about why you moved these out of the `guarantee`?

Reply:
Yep - we have an `IO[Stream[IO, Result]]` here. The `None` was being published after the `IO` completed, but now that happens as soon as the `Stream` is available, which happens before it completes. So instead we need to write the `None` after the `Stream` from `interp.runCode` completes, or we'll terminate the output before the cell is finished.

Reply:
Probably will be easier to draw than explain 😅

Reply:
I guess I was thinking that the equivalent of the guarantee on the IO would for a Stream be something like 
```
.map {
             stream =>
                stream.onFinalize {
                  oq.enqueue1(None) *> symbolTable.drain()
                }
}
```

Reply:
👍 you're right, that is much better. Will change it.

Reply:
So, I tried this out and interestingly it has subtly different semantics.

The current way (appending `Stream.eval`s), those finalization tasks get run *as part of the stream*, while with `onFinalize` they are separate from the stream. This causes weird behavior (i.e. when cancelling a task the `InterruptedException` shows up as a kernel error rather than a `RuntimeError`). So I'm going to leave it this way for now.

Reply:
I wish there was a good way to track TODOs in Github, without making them issues... I've got a lot of "try to clean this up"-type TODOs, but they're not exactly issues...

Reply:
Whoa that's weird! 

Hmm, I guess we could make a TODO milestone and add issues there... Or I guess keep 'em in the codebase and use the IntelliJ TODO browser might be enough.